[13.0][ADD] stock_location_last_inventory_date

carlosserratoro opened the pull request:
This new module adds a stored computed field on locations that stores
the last inventory date for the location, for validated inventories.
This is only computed for leaf locations, not parent ones.

Commit:
[ADD] stock_location_last_inventory_date

This new module adds a stored computed field on locations that stores
the last inventory date for the location, for validated inventories.
This is only computed for leaf locations, not parent ones.

guewen commented on stock_location_last_inventory_date/models/stock_location.py:
```diff
@@ -0,0 +1,44 @@
+                and loc.validated_inventory_ids
+            ):
+                loc.last_inventory_date = loc.validated_inventory_ids.sorted(
+                    lambda inventory: inventory.write_date
+                )[-1].write_date
```
Is there a reason to use `write_date` and not `date`? `write_date` could hypothetically change even on a very old inventory if we call `write()` on it

carlosserratoro replied:
This was a mistake. There is no need to not use ```date```. I'll force-push a correction on this. Thanks!

guewen commented on stock_location_last_inventory_date/models/stock_location.py:
```diff
@@ -0,0 +1,44 @@
+        domain="[('location_ids', 'in', id), ('state', '=', 'done')]",
+    )
+
+    @api.depends("validated_inventory_ids", "validated_inventory_ids.write_date")
```
```suggestion
    @api.depends("validated_inventory_ids", "validated_inventory_ids.date")
```

guewen commented on stock_location_last_inventory_date/models/stock_location.py:
```diff
@@ -0,0 +1,44 @@
+                and loc.validated_inventory_ids
+            ):
+                loc.last_inventory_date = loc.validated_inventory_ids.sorted(
+                    lambda inventory: inventory.write_date
```
```suggestion
                    lambda inventory: inventory.date
```

guewen reviewed (changes requested):
Leftovers of `write_date` to replace by `date`

carlosserratoro committed:
Apply suggestions from code review

Co-authored-by: Guewen Baconnier <guewen.baconnier@camptocamp.com>

sebalix commented on stock_location_last_inventory_date/models/stock_location.py:
```diff
@@ -0,0 +1,44 @@
+        domain="[('location_ids', 'in', id), ('state', '=', 'done')]",
+    )
+
+    @api.depends("validated_inventory_ids", "validated_inventory_ids.date")
```
I would add `usage` and `child_ids` to the dependencies

carlosserratoro replied:
You are right. Thanks! ─ done.

Commit:
[FIX] Add missing fields in the api.depends

Thanks to sebalix for spotting them.

rousseldenis commented on stock_location_last_inventory_date/__manifest__.py:
```diff
@@ -0,0 +1,15 @@
+    "summary": "Show the last inventory date for a leaf location",
+    "version": "13.0.1.0.0",
+    "development_status": "Alpha",
+    "category": "Warehouse Management",
```
Please set to 'Warehouse' to align https://odoo-community.org/shop categories (Odoo standard ones)

carlosserratoro replied:
Done.

rousseldenis reviewed (approved):
LGTM. Minor comment

rousseldenis commented on stock_location_last_inventory_date/models/stock_location.py:
```diff
@@ -0,0 +1,49 @@
+    @api.depends(
+        "validated_inventory_ids",
+        "validated_inventory_ids.date",
+        "validated_inventory_ids.usage",
```
```suggestion
        "validated_inventory_ids.location_id.usage",
```

carlosserratoro replied:
Wait, I'm realising this depends is wrong also because of the child_ids, this should be added also: ```validated_inventory_ids.location_id.child_ids```.

carlosserratoro replied:
I re-wrote the list of depends, there were a couple missing.

rousseldenis reviewed (changes requested):
LGTM. Minor comment

guewen commented:
@carlosserratoro did you try with a normal stock user? I wonder if you won't have an access right issue without `compute_sudo` on the field.

carlosserratoro commented on stock_location_last_inventory_date/models/stock_location.py:
```diff
@@ -0,0 +1,49 @@
+        "validated_inventory_ids",
+        "validated_inventory_ids.date",
+        "validated_inventory_ids.usage",
+        "validated_inventory_ids.child_ids",
```
```suggestion
        "validated_inventory_ids.location_id.child_ids",
```

Commit:
[FIX] stock_buffer_sales_analysis: attending the code review

Several changes were done because of the code review:
* The category of the module is now "Warehouse", in alignment
  with the categories user in OCA (https://odoo-community.org/shop)
* The list of fields in the depends() for the method
  _compute_last_inventory_date() missed several fields, that are
  now listed as dependencies.
* A new unit test has been added to account for the concern of the
  functional needing a compute_sudo in its definition, just in case
  a non privileged stock user was stuck into permission problems.

carlosserratoro commented:
> @carlosserratoro did you try with a normal stock user? I wonder if you won't have an access right issue without `compute_sudo` on the field.

@guewen I tested using unit tests. For indirect recomputations (for instance those that happen when one of the fields in the ```depends``` is changed) it works fine without ```compute_sudo``` because a special user *Sytem* does it. However a call to the compute method with a hardcoded ```.with_user()``` has precedence over the ```compute_sudo```, so it fails, but I think this applies to any computed field in Odoo. To be extra-sure, I can add the ```compute_sudo``` **and** use ```sudo()``` inside the method.

For the moment I've added a new test having to do with permissions (without using ```sudo()```) Can you have a look again, please?

guewen commented:
@carlosserratoro That's good without adding `compute_sudo` or `sudo()` then. I don't think we should expect users able to directly call the compute method. Thanks for checking :)

jgrandguillaume commented:
/ocabot merge nobump
